### PR TITLE
K3s: Use channel server for versions.

### LINKED ISSUE #740: k3s: Switch to fetching versions via channel server
From Slack:
> you should use the channel server instead of GH releases

I believe that refers to https://github.com/rancher/channelserver

- [ ] Revert #744

### PR DESCRIPTION
**Pull request by mook-as**
This changes the version dropdowns to prefer versions coming from the k3s release channel server.

Due to UI reasons (Chrome doesn't support bold / colour / etc. in `<select>`s on macOS), this is done as two sections.

Fixes #740.

### CODE REVIEW

File: src/k8s-engine/k3sHelper.ts [hunk 1]
```diff
@@ -199,12 +210,45 @@ export default class K3sHelper extends events.EventEmitter {
...
+
+            if (a === 'stable' || b === 'stable') {
+              // sort "stable" at the front
+              return a === 'stable' ? -1 : 1;
+            }
+            if (versionRegex.test(a) || versionRegex.test(b)) {
+              return versionRegex.test(a) ? 1 : -1;
+            }
+
+            return a.localeCompare(b);
```
Comment (ericpromislow) [inline src/k8s-engine/k3sHelper.ts:246]: According to the comment at line 232, all the names here must be `"testing"`, in which case you can return 0. Or should the comment be `a word ("stable", "testing", etc.)`?

**Review by ericpromislow — CHANGES_REQUESTED** [review]
Just some typos. Works fine

File: src/k8s-engine/k3sHelper.ts [hunk 2]
```diff
@@ -71,26 +68,34 @@ export default class K3sHelper extends events.EventEmitter {
...
     try {
-      const cacheData: FullVersion[] =
+      const cacheData: (string | { version: string, channels: string[] | undefined })[] =
         JSON.parse(await util.promisify(fs.readFile)(this.cachePath, 'utf-8'));
 
-      for (const versionString of cacheData) {
-        const version = semver.parse(versionString);
+      for (const entry of cacheData) {
+        if (typeof entry === 'string') {
+          // Old styl cache; don't load it, because doing so prevents us from
```
[the remark this answers is not in the document]
Comment (mook-as) [reply src/k8s-engine/k3sHelper.ts]: Fixed in 5a72fb1325fb57c624824a1f24f9624dd867b0b7.

File: src/k8s-engine/k3sHelper.ts [hunk 3]
```diff
@@ -102,10 +107,12 @@ export default class K3sHelper extends events.EventEmitter {
...
   /** Write this.versions into the cache file. */
   protected async writeCache() {
-    const cacheData = JSON.stringify(Object.values(this.versions).map(v => v.raw));
+    const cacheData = Object.values(this.versions).map((entry) => {
+      return { version: entry.version.raw, channels: entry.channels };
+    });
 
     await fs.promises.mkdir(paths.cache, { recursive: true });
-    await fs.promises.writeFile(this.cachePath, cacheData, 'utf-8');
+    await fs.promises.writeFile(this.cachePath, JSON.stringify(cacheData), 'utf-8');
```
[the remark this answers is not in the document]
Comment (mook-as) [reply src/k8s-engine/k3sHelper.ts]: Fixed in 5a72fb1325fb57c624824a1f24f9624dd867b0b7 (though I would have expected a human to run it through `jq`).

File: src/k8s-engine/k3sHelper.ts [hunk 4]
```diff
@@ -173,15 +181,18 @@ export default class K3sHelper extends events.EventEmitter {
         // already seen before for sure.  This is the only situation where we
         // can be sure that we will not find more useful versions.
         console.log(`Found old version ${ version.raw }, stopping.`);
+        console.log(JSON.stringify(this.versions[version.version]), Object.keys(this.versions));
```
[the remark this answers is not in the document]
Comment (mook-as) [reply src/k8s-engine/k3sHelper.ts]: Fixed in 5a72fb1325fb57c624824a1f24f9624dd867b0b7 (`jq` is not relevant here, in contrast to the previous comment).

File: src/k8s-engine/k3sHelper.ts (hunk 1 above)
Comment (mook-as) [reply src/k8s-engine/k3sHelper.ts:248]: Fixed in 5a72fb1325fb57c624824a1f24f9624dd867b0b7 (the second option).

**Review by ericpromislow — APPROVED** [review]
LGTM

**Pull request merged**